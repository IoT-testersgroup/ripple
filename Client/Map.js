import React, { Component } from 'react';
import {
	TouchableOpacity,
	TextInput,
	StyleSheet,
	View,
	Text,
	Dimensions,
} from 'react-native';
import MapView, { Marker } from 'react-native-maps';

//https://snack.expo.io
export default class Map extends Component {
	constructor(props) {
		super(props);
		this.state = {
			// region: {
			// 	latitude: 37.78825,
			// 	longitude: -122.4324,
			// 	latitudeDelta: 0.2,
			// 	longitudeDelta: 0.2,
			// },
			markers: [
				{
					latitude: 37.78825,
					longitude: -122.4324,
					title: 'Foo Place',
					description: '1234 Foo Drive',
				},
			],
		};
	}

	onRegionChange(region) {
		this.setState({ region });
	}

	render() {
		//Error check
		if (this.props.region.latitude === undefined) return <></>;

		return (
			<MapView style={styles.map} region={this.props.region}>
				{this.state.markers.map((marker, index) => {
					// console.log(marker);
					return (
						<MapView.Marker
							key={index}
							coordinate={{
								latitude: marker.latitude,
								longitude: marker.longitude,
							}}
							title={'title'}
							description={'description'}
						/>
					);
				})}
			</MapView>
		);
	}
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		backgroundColor: '#fff',
		alignItems: 'center',
		justifyContent: 'center',
	},
	map: {
		width: Dimensions.get('window').width,
		// height: Dimensions.get('window').height,
		height: '75%',
	},
});

// coordinate={{latitude: marker.latitude,
// 	longitude: marker.longitude}}
