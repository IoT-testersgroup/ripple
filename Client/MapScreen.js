import React from 'react';
import { render } from 'react-dom';
import Map from './Map';
import HeatMap from './HeatMap';
import Slider from '@react-native-community/slider';
import Geocoder from 'react-native-geocoding';
import Ionicons from 'react-native-vector-icons/Ionicons';

import {
	View,
	Text,
	TouchableOpacity,
	StyleSheet,
	TextInput,
} from 'react-native';
Geocoder.init('AIzaSyCeGW7SfROh0vU4a2P16hfYOtv-62nn18M'); // use a valid API key

export default class MapScreen extends React.Component {
	constructor(props) {
		super(props);
		let d = new Date();
		console.log(d.getMonth() + 12);
		this.state = {
			heapmap: false,
			month: d.getMonth() + 12,
			value: d.getMonth() + 12,
			region: {},
			searchString: '',
			allPoints: [],
			filteredPoints: [],
			// aggregatedPoints: [],
		};
		this.getButtonText = this.getButtonText.bind(this);
		this.getMonthByNumber = this.getMonthByNumber.bind(this);
		// this.aggregatePoints = this.aggregatePoints.bind(this);
		this.filterByMonth = this.filterByMonth.bind(this);
	}
	getButtonText() {
		const { heatmap } = this.state;

		if (heatmap) return 'View Default Map';
		else return 'View Heat Map';
	}
	//Get month by number
	getMonthByNumber() {
		// monthnumber %= 12;
		const monthNames = [
			'January',
			'February',
			'March',
			'April',
			'May',
			'June',
			'July',
			'August',
			'September',
			'October',
			'November',
			'December',
		];

		return monthNames[this.state.value % 12]; //if 2
	}
	componentDidMount() {
		this.updateLocation('San Francisco');
	}
	componentDidUpdate() {
		if (this.props.currentTabScreen != 'Incident Map') {
			this.props.navigate('Create Report');
		}
	}
	fetchData(region) {
		// console.log('fetching data for: ');
		// console.log(region);
		fetch('https://ripple506.herokuapp.com/getReportByLocation', {
			method: 'POST',
			headers: {
				'Accept': '*/*',
				'Connection': 'Keep-Alive',
				'Content-Type': 'application/json',
			},

			body: JSON.stringify({
				'Latitude': region.latitude,
				'Longitude': region.longitude,
				'LongitudeDelta': region.longitudeDelta,
				'LatitudeDelta': region.latitudeDelta,
			}),
		})
			.then((response) => response.json())

			.then(async (json) => {
				// console.log('DATA: ');
				// console.log(json.Data);
				var allPoints = json.Data.map(function (object) {
					var newObject = Object.assign({}, object);
					newObject.weight = 10;
					let tempLat = newObject.Latitude;
					let tempLong = newObject.Longitude;
					delete newObject.Latitude;
					delete newObject.Longitude;
					newObject.latitude = tempLat;
					newObject.longitude = tempLong;
					newObject.title = newObject.Category;
					newObject.description = newObject.Details;
					delete newObject.Category;
					delete newObject.Details;
					return newObject;
				});

				this.setState({ allPoints, filteredPoints: allPoints });
				// console.log(json);
				if (json.Status) {
				}
			});
	}
	filterByMonth(value) {
		const { allPoints, month } = this.state;
		this.setState({ value });
		const filteredPoints = allPoints.filter(function (point) {
			return parseInt(point.Month) <= value;
		});
		// console.log(filteredPoints);
		// allPoints.filter(parseInt(point.Month) <= month);
		this.setState({ filteredPoints });
	}
	// aggregatePoints() {
	// 	const { allPoints } = this.state;
	// 	let aggregatedPoints = [];
	// 	for (let i = 0; i < allPoints.length; i++) {
	// 		let found = false;
	// 		for (let j = 0; j < aggregatedPoints.length; j++) {
	// 			if (
	// 				aggregatedPoints[j].latitude === allPoints[i].latitude &&
	// 				aggregatedPoints[j].longitude === allPoints[i].longitude
	// 			) {
	// 				found = true;
	// 				aggregatedPoints[j].description += allPoints[i].description;
	// 				break;
	// 			}
	// 		}
	// 		if (!found) aggregatedPoints.push(allPoints[i]);
	// 	}
	// 	this.setState({ aggregatedPoints });
	// }
	updateLocation(searchString) {
		console.log('Searching for' + searchString);
		Geocoder.from(searchString)
			.then((json) => {
				var location = json.results[0].geometry.location;
				let newRegion = {};
				newRegion.latitude = Number(location.lat.toFixed(4));
				newRegion.longitude = Number(location.lng.toFixed(4));
				newRegion.latitudeDelta = 0.2;
				newRegion.longitudeDelta = 0.2;
				this.setState({ region: newRegion });
				this.fetchData(newRegion);
			})
			.catch((error) => console.warn(error));
	}
	render() {
		// Search by address

		// console.log(this.state.region);
		const { heatmap, month, allPoints, filteredPoints, region } = this.state;
		const map =
			heatmap === true ? (
				<HeatMap region={region} points={filteredPoints} />
			) : (
				<Map region={region} points={filteredPoints} />
			);
		return (
			<View>
				<Text style={{ textAlign: 'center', fontSize: 30 }}>
					{this.getMonthByNumber()}
				</Text>
				<Slider
					className='slider'
					style={{ width: '90%', marginLeft: '5%', height: 40 }}
					minimumValue={month - 11}
					step={1}
					value={month}
					maximumValue={month}
					onValueChange={(value) => this.filterByMonth(value)}
					minimumTrackTintColor='#FFFFFF'
					maximumTrackTintColor='#000000'
				/>
				{map}
				<View style={styles.searchSection}>
					<Ionicons
						style={styles.searchIcon}
						name='search'
						size={20}
						color='#000'
						onPress={() => this.updateLocation(this.state.searchString)}
					/>
					<TextInput
						style={styles.input}
						placeholder='Enter a Location'
						onChangeText={(searchString) => {
							this.setState({ searchString });
						}}
						onSubmitEditing={() => this.updateLocation(this.state.searchString)}
					/>
				</View>
				<TouchableOpacity
					style={styles.button}
					// onPress={() => this.setState({ heatmap: !heatmap })}>
					onPress={() =>
						this.setState(
							{
								heatmap: !this.state.heatmap,
							} /*, alert('Heatmap not implemented')*/
						)
					}>
					<Text style={{ fontSize: 18, color: 'white' }}>
						{this.getButtonText()}
					</Text>
				</TouchableOpacity>
			</View>
		);
	}
}
const styles = StyleSheet.create({
	button: {
		justifyContent: 'center',
		alignItems: 'center', // Centered horizontally
		justifyContent: 'center', //Centered vertically
		height: '5%',
		width: 200,
		margin: 5,
		backgroundColor: 'black',
		padding: 5,
		borderRadius: 40,
		borderWidth: 1,
	},
	input: {
		padding: 10,
		margin: 5,
		height: 40,
		width: '80%',
		borderColor: 'black',
		borderWidth: 1,
	},
	searchSection: {
		flex: 1,
		flexDirection: 'row',
		justifyContent: 'center',
		alignItems: 'center',
		backgroundColor: '#fff',
	},
	searchIcon: {
		padding: 10,
	},
	input: {
		flex: 1,
		paddingTop: 10,
		paddingRight: 10,
		paddingBottom: 10,
		paddingLeft: 0,
		backgroundColor: '#fff',
		color: '#424242',
	},
});
