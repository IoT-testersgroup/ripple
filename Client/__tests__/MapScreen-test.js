import React from 'react';
import { renderer, mount } from 'react-test-renderer';
import MapScreen from '../MapScreen.js';
import Map from '../Map.js';

jest.useFakeTimers();

//Need to use await because componentDidMount uses fetch
describe('<MapScreen/>', () => {
	it('MapScreen originally has 4 children', async () => {
		const tree = await renderer.create(<MapScreen />).toJSON();
		console.log(tree.children);
		expect(tree.children.length).toBe(4);
		expect(tree).toMatchSnapshot();
	});
	it('Correct Month displayed originally', async () => {
		const tree = await renderer.create(<MapScreen />).toJSON();
		const date = new Date();
		const month = date.toLocaleString('default', { month: 'long' });
		expect(tree.children[0].children[0]).toBe(month);
	});

	//On button change, components are correct

	//On slide, the month is accurate

	//On slide, the filtering is accurate
});

describe('<Map/>', () => {
	it('Map has one point for every single prop', async () => {});

	//Map
});

//Heatmap has one point for every single prop
