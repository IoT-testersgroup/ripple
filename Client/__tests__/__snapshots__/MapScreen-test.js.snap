// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MapScreen/> MapScreen originally has 4 children 1`] = `
<View>
  <Text
    style={
      Object {
        "fontSize": 30,
        "textAlign": "center",
      }
    }
  >
    April
  </Text>
  <RNCSlider
    className="slider"
    disabled={false}
    enabled={true}
    inverted={false}
    maximumTrackTintColor="#000000"
    maximumValue={15}
    minimumTrackTintColor="#FFFFFF"
    minimumValue={4}
    onChange={[Function]}
    onRNCSliderSlidingComplete={null}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={1}
    style={
      Array [
        Object {
          "height": 40,
        },
        Object {
          "height": 40,
          "marginLeft": "5%",
          "width": "90%",
        },
      ]
    }
    thumbImage={
      Object {
        "__packager_asset": true,
        "height": 1,
        "scale": 1,
        "uri": "uri",
        "width": 1,
      }
    }
    value={15}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Enter a Location"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "color": "#424242",
          "flex": 1,
          "paddingBottom": 10,
          "paddingLeft": 0,
          "paddingRight": 10,
          "paddingTop": 10,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "black",
        "borderRadius": 40,
        "borderWidth": 1,
        "height": "5%",
        "justifyContent": "center",
        "margin": 5,
        "opacity": 1,
        "padding": 5,
        "width": 200,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
        }
      }
    >
      View Heat Map
    </Text>
  </View>
</View>
`;
